plugins/iamauth.js

```diff
@@ -29,165 +29,77 @@ class IAMPlugin extends BasePlugin {
       throw new Error('Missing IAM API key from configuration');
     }
 
-    // token service retry configuration
     cfg = Object.assign({
-      retryDelayMultiplier: 2,
-      retryInitialDelayMsecs: 500
+      autoRenew: true,
+      iamTokenUrl: 'https://iam.cloud.ibm.com/identity/token',
+      retryDelayMsecs: 1000
     }, cfg);
 
     super(client, serverUrl, cfg);
 
-    this.currentIamApiKey = null;
-    this.baseUrl = cfg.baseUrl || null;
-    this.cookieJar = null;
-    this.tokenUrl = cfg.iamTokenUrl || 'https://iam.cloud.ibm.com/identity/token';
-    this.refreshRequired = true;
+    let sessionUrl = new u.URL(serverUrl);
+    sessionUrl.pathname = '/_iam_session';
+
+    this._jar = request.jar();
+
+    this._tokenManager = new IAMTokenManager(
+      client,
+      this._jar,
+      u.format(sessionUrl, {auth: false}),
+      cfg.iamTokenUrl,
+      cfg.iamApiKey,
+      cfg.iamClientId,
+      cfg.iamClientSecret
+    );
+
+    if (cfg.autoRenew) {
+      this._tokenManager.startAutoRenew();
+    }
   }
 
   onRequest(state, req, callback) {
     var self = this;
 
-    if (self._cfg.iamApiKey !== self.currentIamApiKey) {
-      debug('New IAM API key identified.');
-      self.currentIamApiKey = self._cfg.iamApiKey;
-      state.stash.newApiKey = true;
-    } else if (!self.refreshRequired) {
-      if (self.baseUrl && self.cookieJar.getCookies(self.baseUrl, {expire: true}).length === 0) {
-        debug('There are no valid session cookies in the jar. Requesting IAM session refresh...');
-      } else {
-        req.jar = self.cookieJar; // add jar
-        return callback(state);
-      }
-    }
-
-    req.url = req.uri || req.url;
-    delete req.uri;
+    req.jar = self._jar;
 
-    if (self.baseUrl === null) {
-      var parsed = u.parse(req.url);
-      self.baseUrl = u.format({
-        protocol: parsed.protocol,
-        host: parsed.host,
-        port: parsed.port
-      });
-    }
+    req.uri = req.uri || req.url;
+    delete req.url;
+    req.uri = u.format(new u.URL(req.uri), {auth: false});
 
-    self.refreshCookie(self._cfg, state, function(error) {
-      if (error) {
-        debug(error.message);
-        if (state.attempt < state.maxAttempt) {
-          state.retry = true;
-          if (state.attempt === 1) {
-            state.retryDelayMsecs = self._cfg.retryInitialDelayMsecs;
-          } else {
-            state.retryDelayMsecs *= self._cfg.retryDelayMultiplier;
-          }
+    self._tokenManager.renew().then(() => {
+      callback(state);
+    }).catch((error) => {
+      debug(error);
+      if (state.attempt < state.maxAttempt) {
+        state.retry = true;
+        let iamResponse = error.response[0];
```

Comment:
I think I'm still mis-understanding something here, but it should be impossible to get an error from the IAM token server followed by an error from `_iam_session` because if there was an error from the token server we shouldn't even make a request to `_iam_session`.
Is the reasoning behind the array just to help you unwrap whether the error was from the first request or the second when you've lost other associated context? Can we just use a specific error type instead?

Reply:
@smithsz any thoughts on this

Reply:
As discussed, I'll return only the last response in the array to avoid any confusion.